File: 0127-word-ladder/0127-word-ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        if(beginWord.length() != endWord.length()) return 0;

        Set<String> words = new HashSet<>(wordList);
        if(!words.contains(endWord)) return 0;

        int cnt = 1;
        Queue<String> q = new LinkedList<>();
        q.offer(beginWord);
        words.remove(beginWord);

        while(!q.isEmpty()){
            int size = q.size();

            for(int i = 0; i < size; i++){

                String currWord = q.poll();
                if(currWord.equals(endWord)) return cnt;
                char[] wordArray = currWord.toCharArray();

                for(int j = 0; j < wordArray.length; j++){
                    char orignalChar = wordArray[j];
                    
                    for(char c = 'a'; c <= 'z'; c++){
                        if(c == orignalChar) continue;
                        wordArray[j] = c;
                        String nextWord = new String(wordArray);
                        if(words.contains(nextWord)){
                            q.offer(nextWord);
                            words.remove(nextWord);
                        }
                    }

                    wordArray[j] = orignalChar;
                }
            }

            cnt++;
        }

        return 0;
    }
}

File: 0127-word-ladder/0127-word-ladder.py
class Solution(object):
    def ladderLength(self, beginWord, endWord, wordList):
        if len(beginWord) != len(endWord):
            return 0
        
        word_set = set(wordList)

        if endWord not in word_set:
            return 0

        count = self.helper(beginWord, endWord, word_set)

        if count == 0:
            return 0
        
        return count
    
    def helper(self, beginWord, endWord, word_set):
        cnt = 1
        queue = deque([beginWord])
        while queue:
            size = len(queue)
            for i in range(size):
                curr_word = queue.popleft()
                if curr_word == endWord:
                    return cnt
                
                wordArray = list(curr_word)
                for j in range(len(wordArray)):
                    curr_char = wordArray[j]
                    for c in range(ord('a'), ord('z') + 1):
                        if chr(c) == curr_char:
                            continue
                        wordArray[j] = chr(c)
                        new_word = ''.join(wordArray)
                        if new_word in word_set:
                            queue.append(new_word)
                            word_set.remove(new_word)
                    
                    wordArray[j] = curr_char
                
            cnt += 1
        
        return 0


        

        